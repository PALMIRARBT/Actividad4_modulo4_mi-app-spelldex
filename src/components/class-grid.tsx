import c from "classnames";
import classesJson from "src/data/classes.json";

import { ClassGridItem } from "./class-item";
import { useRef } from "react";

import type { CharacterClass, ClassId } from "src/models/character-class";

import styles from "./class-grid.module.css";

type Props = {
  selectedClass?: ClassId;
  background?: boolean;
  onClick: (c: ClassId | undefined) => void;
  highlight: (c: ClassId | undefined) => void;
};

const classes = classesJson as CharacterClass[];

const KEYBOARD_KEYS = [
  "ArrowRight",
  "ArrowLeft",
  "ArrowUp",
  "ArrowDown",
  "Escape",
  "Backspace",
  "Tab",
];

export function ClassGrid({
  selectedClass,
  background,
  highlight,
  onClick,
}: Props) {
  const ref = useRef<HTMLDivElement>(null);
  const itemButtonRefs = useRef<(HTMLButtonElement | null)[]>([]);

  const items = selectedClass
    ? classes.filter((cls) => cls.slug === selectedClass)
    : classes;

  const currentHighlight = () => {
    return selectedClass
      ? items.findIndex((cls) => cls.slug === selectedClass)
      : Array.from(itemButtonRefs.current).findIndex(
          (ref) => ref === document.activeElement
        );
  };

  const keyDown = (event: React.KeyboardEvent) => {
    if (!KEYBOARD_KEYS.includes(event.key)) {
      return;
    }

    // Permitir ESC y Backspace en cualquier momento
    if (event.key === "Escape" || event.key === "Backspace") {
      event.preventDefault();
      onClick(undefined);
      return;
    }

    // El resto solo si el foco est√° en un classCell
    if (!document.activeElement?.classList.contains(styles.classCell)) {
      return;
    }

    const currentIndex = currentHighlight();
    if (currentIndex < 0) {
      return;
    }

    let nextIndex = currentIndex;
    switch (event.key) {
      case "ArrowRight":
      case "ArrowDown":
        nextIndex = Math.min(items.length - 1, currentIndex + 1);
        event.preventDefault();
        itemButtonRefs.current[nextIndex]?.focus();
        break;
      case "ArrowLeft":
      case "ArrowUp":
        nextIndex = Math.max(0, currentIndex - 1);
        event.preventDefault();
        itemButtonRefs.current[nextIndex]?.focus();
        break;
      case "Tab":
        if (!event.shiftKey) {
          nextIndex = (currentIndex + 1) % items.length;
        } else {
          nextIndex = (currentIndex - 1 + items.length) % items.length;
        }
        event.preventDefault();
        itemButtonRefs.current[nextIndex]?.focus();
        break;
      default:
        break;
    }
  };

  return (
    <section
      ref={ref}
      className={c(styles.classGrid, background && styles.background)}
      aria-label="Class Grid"
      onKeyDown={keyDown}
      role="grid"
    >
      {items.map((cls, index) => (
        <ClassGridItem
          key={cls.slug}
          name={cls.name}
          classId={cls.slug}
          selected={selectedClass === cls.slug}
          highlight={highlight}
          onClick={() => onClick(cls.slug)}
          ref={(el) => {
            if (el) {
              itemButtonRefs.current[index] = el;
              if (selectedClass === cls.slug) {
                el.focus();
              }
            } else {
              itemButtonRefs.current[index] = null;
            }

            return () => {
              itemButtonRefs.current[index] = null;
            };
          }}
        />
      ))}
    </section>
  );
}
