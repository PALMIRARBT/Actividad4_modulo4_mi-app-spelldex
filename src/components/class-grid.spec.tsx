import { render, screen } from "@testing-library/react";
import { ClassGrid } from "./class-grid";

describe("<ClassGrid />", () => {
  it('renders the grid: match snapshot', () => {
    render(<ClassGrid onClick={() => {}} highlight={() => {}} />);

    const section = screen.getByRole("grid", {
      name: "Class Grid",
    });
    expect(section).toBeInTheDocument();
    expect(section).toMatchSnapshot();
  });
});
