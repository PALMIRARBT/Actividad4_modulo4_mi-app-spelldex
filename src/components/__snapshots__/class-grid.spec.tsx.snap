// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClassGrid /> renders the grid: match snapshot 1`] = `
<section
  aria-label="Class Grid"
  role="grid"
>
  <button
    aria-label="Bard"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <img
        alt="Bard"
        src="src/assets/icons/classes/bard.png"
      />
    </div>
    <span>
      Bard
    </span>
  </button>
  <button
    aria-label="Cleric"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <img
        alt="Cleric"
        src="src/assets/icons/classes/cleric.png"
      />
    </div>
    <span>
      Cleric
    </span>
  </button>
  <button
    aria-label="Druid"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <img
        alt="Druid"
        src="src/assets/icons/classes/druid.png"
      />
    </div>
    <span>
      Druid
    </span>
  </button>
  <button
    aria-label="Sorcerer"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <img
        alt="Sorcerer"
        src="src/assets/icons/classes/sorcerer.png"
      />
    </div>
    <span>
      Sorcerer
    </span>
  </button>
  <button
    aria-label="Warlock"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <img
        alt="Warlock"
        src="src/assets/icons/classes/warlock.png"
      />
    </div>
    <span>
      Warlock
    </span>
  </button>
  <button
    aria-label="Wizard"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <div
        class=""
      />
      <img
        alt="Wizard"
        src="src/assets/icons/classes/wizard.png"
      />
    </div>
    <span>
      Wizard
    </span>
  </button>
</section>
`;
